src/Utilities/functions.jl

```diff
@@ -3409,3 +3409,87 @@ function MA.promote_operation(
 end
 
 Base.conj(f::Union{MOI.VariableIndex,MOI.VectorOfVariables}) = f
+
+## Matrix operations
+
+function promote_operation(
+    ::typeof(*),
+    ::Type{T},
+    ::Type{<:Diagonal{T}},
+    ::Type{F},
+) where {T,F}
+    S = scalar_type(F)
+    U = promote_operation(*, T, T, S)
+    return promote_operation(vcat, T, U)
+end
+
+function operate_term(::typeof(*), D::Diagonal, term::MOI.VectorAffineTerm)
+    return MOI.VectorAffineTerm(
+        term.output_index,
+        operate_term(*, D.diag[term.output_index], term.scalar_term),
+    )
+end
+
+function operate_term(::typeof(*), D::Diagonal, term::MOI.VectorQuadraticTerm)
+    return MOI.VectorQuadraticTerm(
```

Comment:
If we're going to add these, they need tests?

Reply:
At the moment, yes